### PR TITLE
feat: Rotate pods feature

### LINKED ISSUE #3276: Feature: Rotate pods option in Deployment pipeline
### 🔖 Feature description

There should be a checkbox on the deployment pipeline like `Rotate pods` selecting which will forcefully rotate the pods even though there's no change in pod spec.

### 🎤 Pitch

There are usecases where the pods need to be rotated and Devtron does not gives a mechanism out of the box to do that.
The cases may be such as:
- A common configmap value has changed and the pods need to be rotated so that the new values can be loaded.

### ✌️ Solution

There should be a checkbox on the deployment pipeline like `Rotate pods` selecting which will forcefully rotate the pods even though there's no change in pod spec.
This can be done by injecting an environment variable in the pod specs like, the value of which can be autogenerated:
`RELEASE_HASH: ahdklyuu569Ghs54nmk6`

### 🔄️ Alternative

As an alternative, a small change in the pod spec like resources.limit is done and the deployment is re-triggered.

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/devtron-labs/devtron/blob/main/CODE_OF_CONDUCT.md)

AB#2735

### PR DESCRIPTION
**Pull request by kripanshdevtron**
# Description

Fixes #3276 

## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist:

* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have performed a self-review of my own code.
* [ ] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/pipeline/WorkflowDagExecutor.go [hunk 1]
```diff
@@ -1268,6 +1274,39 @@ type StopDeploymentGroupRequest struct {
...
+	}
+	rotatePodRequest := &k8s.RotatePodRequest{
+		ClusterId: environment.ClusterId,
+		Resources: resourceIdentifiers,
+	}
+	response, err := impl.k8sApplicationService.RotatePods(ctx, rotatePodRequest)
+	if err != nil {
+		return nil, err
+	}
+	//TODO KB: make entry in cd workflow runner
```
Comment (kartik-579) [inline pkg/pipeline/WorkflowDagExecutor.go:1306]: is this to be done in this story only?

File: util/k8s/k8sApplicationService.go [hunk 2]
```diff
@@ -712,6 +715,73 @@ func (impl *K8sApplicationServiceImpl) GetResourceList(ctx context.Context, toke
...
+		groupVersionKind := resourceIdentifier.GroupVersionKind
+		resourceKind := groupVersionKind.Kind
+		// validate one of deployment, statefulset, daemonSet, Rollout
+		if resourceKind != kube.DeploymentKind && resourceKind != kube.StatefulSetKind && resourceKind != kube.DaemonSetKind && resourceKind != util.K8sClusterResourceRolloutKind {
+			impl.logger.Errorf("restarting not supported for kind %s name %s", resourceKind, resourceIdentifier.Name)
+			containsError = true
+			resourceResponse.ErrorResponse = "restarting not supported"
+		} else {
+			activitySnapshot := time.Now().Format(time.RFC3339)
+			data := fmt.Sprintf(`{"metadata": {"annotations": {"devtron.ai/restartedAt": "%s"}},"spec": {"template": {"metadata": {"annotations": {"devtron.ai/activity": "%s"}}}}}`, activitySnapshot, activitySnapshot)
```
Comment (kartik-579) [inline util/k8s/k8sApplicationService.go:762]: can make const

File: util/k8s/k8sApplicationService.go [hunk 3]
```diff
@@ -712,6 +715,73 @@ func (impl *K8sApplicationServiceImpl) GetResourceList(ctx context.Context, toke
...
+		resourceResponse := &RotatePodResourceResponse{
+			ResourceIdentifier: resourceIdentifier,
+		}
+		groupVersionKind := resourceIdentifier.GroupVersionKind
+		resourceKind := groupVersionKind.Kind
+		// validate one of deployment, statefulset, daemonSet, Rollout
+		if resourceKind != kube.DeploymentKind && resourceKind != kube.StatefulSetKind && resourceKind != kube.DaemonSetKind && resourceKind != util.K8sClusterResourceRolloutKind {
+			impl.logger.Errorf("restarting not supported for kind %s name %s", resourceKind, resourceIdentifier.Name)
+			containsError = true
+			resourceResponse.ErrorResponse = "restarting not supported"
```
Comment (kartik-579) [inline util/k8s/k8sApplicationService.go:759]: can make const

**Review by kartik-579 — CHANGES_REQUESTED** [review]

File: pkg/pipeline/WorkflowDagExecutor.go (hunk 1 above)
Comment (kripanshdevtron) [reply pkg/pipeline/WorkflowDagExecutor.go:1306]: this would be done in phase 2 of it.

File: util/k8s/k8sApplicationService.go (hunk 2 above)
Comment (kripanshdevtron) [reply util/k8s/k8sApplicationService.go:762]: I don't think this is required.

**Review by kartik-579 — APPROVED** [review]

**Pull request merged**